picard/util/__init__.py

```diff
@@ -1183,4 +1184,10 @@ def detect_unicode_encoding(path):
         for bom, encoding in ENCODING_BOMS.items():
```

Comment:
If we're going to use `chardet` this code is likely a bit redundant.

Reply:
actually chardet doesn't seem to differentiate well between various `utf-16` and `utf-32` encodings. The BOM based test function failed as chardet considered all 4 to be of type utf-8:

            b'\xff\xfe': 'utf-16-le',
            b'\xfe\xff': 'utf-16-be',
            b'\00\00\xff\xfe': 'utf-32-le',
            b'\00\00\xfe\xff': 'utf-32-be',
            
   however it worked correctly for the log file

Reply:
> Also this page says that 'utf-32-le' has a BOM of `b'\xff\xfe\x00\x00'`. That said utf-32 is pretty rare as are the other encodings in the table on the Wikipedia page.

I think it's just a bug and the existing entry of `b'\00\00\xff\xfe'` needs to be updated.

Reply:
> If we're going to use chardet this code is likely a bit redundant.

I probably would keep this. If we get a BOM I think it is fair to assume the corresponding encoding, and we avoid the messy stuff that character encoding guessing is. And the code is simple enough.

Reply:
I tried chardet alternative and the best one `chardet_normalizer` also struggles (more) with above 4 types. It also doesn't have give any confidence value, therefore [checking confidence](https://github.com/metabrainz/picard/pull/2372#discussion_r1521161792) won't be possible with it. However it is considered faster and more accurate model generally. Which package shall I go with for now ?

Reply:
> According to [Wikipedia](https://en.wikipedia.org/wiki/Byte_order_mark) (yes, I know that this is not necessarily definitive or correct), we should also check for `b'\xef\xbb\xbf'` as 'utf-8'.
> 
> Also this page says that 'utf-32-le' has a BOM of `b'\xff\xfe\x00\x00'`. That said utf-32 is pretty rare as are the other encodings in the table on the Wikipedia page.

- utf-8 checking is also available in current test function and it is detected perfectly.
- Also considering `chardet` struggles with these 4 encodings detection, so should I totally remove BOM based detection or keep it for now ?

Reply:
If we give up on checking confidence, we could support both libs, if we stick to https://charset-normalizer.readthedocs.io/en/latest/user/getstarted.html#backward-compatibility style.

Imho, it could be good in case one of those goes unmaintained.

I would choose `chardet_normalizer` over `chardet`, because it's smaller, faster, and detect more charsets.
@phw what's your take on this?

Reply:
I agree, supporting both with the compatbility API, but preferring chardet_normalizer. Which means we should try to import chardet_normalizer and if that fail try chardet. But in requirements-*.txt files we use chardet_normalizer.

Makes sense?

> Also considering chardet struggles with these 4 encodings detection, so should I totally remove BOM based detection or keep it for now ?

I would still opt for keeping it. @zas? We could even make chardet/chardet_normalizer completely optional that way

Reply:
> > 
> 
> I would still opt for keeping it. @zas? We could even make chardet/chardet_normalizer completely optional that way

I agree.